xdsl/dialects/riscv.py

```diff
@@ -498,24 +479,24 @@ class RdImmIntegerOperation(IRDLOperation, RISCVInstruction, ABC):
     immediate operand (e.g. U-Type and J-Type instructions in the RISC-V spec).
     """
 
-    rd: OpResult = result_def(RegisterType)
+    rd: OpResult = result_def(RegisterAttr)
     immediate: AnyIntegerAttr | LabelAttr = attr_def(AnyIntegerAttr | LabelAttr)
 
     def __init__(
         self,
         immediate: int | AnyIntegerAttr | str | LabelAttr,
         *,
-        rd: RegisterType | Register | None = None,
+        rd: RegisterAttr | str | None = None,
```

Comment:
Use of `str` may lost errors that can catch by type checkers in previous implementation. Can you revert it?

Reply:
I still don't see the benefit of the type checking here.

Reply:
If a client wants to give a bogus string then so be it, I don't see the problem. Almost all the time they're expected to use one of the `RegisterAttr` instances in `Registers` or None. I'd rather remove `str` than add back `Register` TBH, but I also don't see it as a serious potential risk.

Reply:
BTW this is pretty much equivalent to allowing int instead of an IntegerAttr, there can be type mismatches, but we allow it for convenience's sake

Reply:
I don't think so. At least `IntegerAttr` has a verifier.

Reply:
So does this class

Reply:
I just went over the codebase and the one location where this matters is `xdsl/transforms/lower_riscv_func.py`:

```diff
            ops.append(
                riscv.MVOp(
                    arg,
-                    rd=riscv.Register(f"a{i}"),
+                    rd=f"a{i}",
                )
            )
```

as well as similar changes in this very same file. Looking into the implementation of Register, I see that Register also did not verify the name of the Register that was passed in its initializer. Hence, the current change seems to just implement what was already there. If there is a more typesafe way to write `lower_riscv_func.py`, we could explore that in the future, but I would suggest to remain permissive for now. In the spirit of reducing feature-creep in our PRs, maybe lets get this one in and then get the custom printing in next?